[WK2] Generalize IPC encoding, decoding of easily-initializable types

zdobersek opened the pull request:
#### 5b3f76435b5ee0346be3bfb418a44f1e9b1228a5
<pre>
[WK2] Generalize IPC encoding, decoding of easily-initializable types
<a href="https://bugs.webkit.org/show_bug.cgi?id=246505">https://bugs.webkit.org/show_bug.cgi?id=246505</a>

Reviewed by NOBODY (OOPS!).

Currently still a RFC-patch.

Minimize the IPC encoding and decoding methods for types whose current
decoding operations are relatively trivial and can be constructed through
list-initialization.

For each such type, a tuple of pointer values to each member variable is
defined. During encoding, this tuple is used to access each member variable
that&apos;s to be encoded. During decoding, the type for each member variable is
deduced from the tuple&apos;s elements, and when successful, the resulting tuple
is expanded into list-initialization-based constructor used for the freshly
decoded object.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::constructObjectFromTupleImpl):
(WTF::constructObjectFromTuple):
* Source/WebKit/GPUProcess/GPUProcessCreationParameters.cpp:
(WebKit::GPUProcessCreationParameters::encode const):
(WebKit::GPUProcessCreationParameters::decode):
* Source/WebKit/GPUProcess/GPUProcessCreationParameters.h:
(WebKit::GPUProcessCreationParameters::ipcData):
* Source/WebKit/GPUProcess/GPUProcessSessionParameters.h:
(WebKit::GPUProcessSessionParameters::ipcData):
(WebKit::GPUProcessSessionParameters::encode const):
(WebKit::GPUProcessSessionParameters::decode):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackendCreationParameters.h:
(WebKit::RemoteRenderingBackendCreationParameters::ipcData):
(WebKit::RemoteRenderingBackendCreationParameters::encode const):
(WebKit::RemoteRenderingBackendCreationParameters::decode):
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.h:
(WebKit::InitializationSegmentInfo::TrackInformation::ipcData):
(WebKit::InitializationSegmentInfo::TrackInformation::encode const):
(WebKit::InitializationSegmentInfo::TrackInformation::decode):
(WebKit::InitializationSegmentInfo::ipcData):
(WebKit::InitializationSegmentInfo::encode const):
(WebKit::InitializationSegmentInfo::decode):
* Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
(WebKit::MediaDescriptionInfo::ipcData):
(WebKit::MediaDescriptionInfo::encode const):
(WebKit::MediaDescriptionInfo::decode):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.h:
(WebKit::RemoteMediaPlayerProxyConfiguration::ipcData):
(WebKit::RemoteMediaPlayerProxyConfiguration::encode const):
(WebKit::RemoteMediaPlayerProxyConfiguration::decode):
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.h:
(WebKit::TextTrackPrivateRemoteConfiguration::ipcData):
(WebKit::TextTrackPrivateRemoteConfiguration::encode const):
(WebKit::TextTrackPrivateRemoteConfiguration::decode):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyDataForSpecificFirstParty::encode const):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyDataForSpecificFirstParty::decode):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyData::encode const):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyData::decode):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
(WebKit::NetworkSessionCreationParameters::ipcData):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeIPCData):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/WebProcess/GPU/media/MediaOverridesForTesting.h:
(WebKit::MediaOverridesForTesting::ipcData):
(WebKit::MediaOverridesForTesting::encode const):
(WebKit::MediaOverridesForTesting::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.h:
(WebKit::RemoteAudioSessionConfiguration::ipcData):
(WebKit::RemoteAudioSessionConfiguration::encode const):
(WebKit::RemoteAudioSessionConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMConfiguration.h:
(WebKit::RemoteCDMConfiguration::ipcData):
(WebKit::RemoteCDMConfiguration::encode const):
(WebKit::RemoteCDMConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceConfiguration.h:
(WebKit::RemoteCDMInstanceConfiguration::ipcData):
(WebKit::RemoteCDMInstanceConfiguration::encode const):
(WebKit::RemoteCDMInstanceConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.h:
(WebKit::RemoteMediaPlayerConfiguration::ipcData):
(WebKit::RemoteMediaPlayerConfiguration::encode const):
(WebKit::RemoteMediaPlayerConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.h:
(WebKit::RemoteMediaPlayerState::ipcData):
(WebKit::RemoteMediaPlayerState::encode const):
(WebKit::RemoteMediaPlayerState::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
(WebKit::RemoteVideoFrameProxy::Properties::encode const): Deleted.
(WebKit::RemoteVideoFrameProxy::Properties::decode): Deleted.
* Source/WebKit/WebProcess/Network/NetworkProcessConnectionInfo.h:
(WebKit::NetworkProcessConnectionInfo::ipcData):
(WebKit::NetworkProcessConnectionInfo::encode const):
(WebKit::NetworkProcessConnectionInfo::decode):
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.h:
(WebKit::WebSpeechSynthesisVoice::ipcData):
(WebKit::WebSpeechSynthesisVoice::encode const):
(WebKit::WebSpeechSynthesisVoice::decode):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b3f76435b5ee0346be3bfb418a44f1e9b1228a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92849 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102576 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2064 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30403 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85252 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98729 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1409 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79339 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28331 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83049 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71447 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36807 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16952 "Found 1 new test failure: webanimations/accelerated-web-animation-with-single-interval-and-easing-y-axis-above-1.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34607 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18142 "Found 1 new test failure: compositing/backing/backing-store-attachment-animating-outside-viewport.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40742 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37320 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUProcessSessionParameters.h:
```diff
@@ -41,48 +41,28 @@ struct GPUProcessSessionParameters {
+    }
+
+    template<class Encoder>
+    void encode(Encoder& encoder) const
```
I'm not entirely fond of adding yet another method for encoding and decoding.
I'd be interested more in approach of 
```
template<class Endecoder>
void endecode(Endecoder& endecoder) DOSOMETHINGWITHCONST {
   endecoder(mediaCacheDirectory)(mediaCacheDirectorySandboxExtensionHandle)(mediaKeysStorageDirectory)(and-so-on);
}
```

Alternatively call `ipcData()` as `codedFields()` or something, and have the Encoder/Decoder templates detect that method only, so that the extra encode/decode boilerplate is not needed.

kkinnunen-apple replied:
There is also some work being done about generating the encoders/decoders, but unfortunately I haven't followed that one at all..

zdobersek committed:
[WK2] Generalize IPC encoding, decoding of easily-initializable types
https://bugs.webkit.org/show_bug.cgi?id=246505

Reviewed by NOBODY (OOPS!).

Currently still a RFC-patch.

Minimize the IPC encoding and decoding methods for types whose current
decoding operations are relatively trivial and can be constructed through
list-initialization.

For each such type, a tuple of pointer values to each member variable is
defined. During encoding, this tuple is used to access each member variable
that's to be encoded. During decoding, the type for each member variable is
deduced from the tuple's elements, and when successful, the resulting tuple
is expanded into list-initialization-based constructor used for the freshly
decoded object.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::constructObjectFromTupleImpl):
(WTF::constructObjectFromTuple):
* Source/WebKit/GPUProcess/GPUProcessCreationParameters.cpp:
(WebKit::GPUProcessCreationParameters::encode const):
(WebKit::GPUProcessCreationParameters::decode):
* Source/WebKit/GPUProcess/GPUProcessCreationParameters.h:
(WebKit::GPUProcessCreationParameters::ipcData):
* Source/WebKit/GPUProcess/GPUProcessSessionParameters.h:
(WebKit::GPUProcessSessionParameters::ipcData):
(WebKit::GPUProcessSessionParameters::encode const):
(WebKit::GPUProcessSessionParameters::decode):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackendCreationParameters.h:
(WebKit::RemoteRenderingBackendCreationParameters::ipcData):
(WebKit::RemoteRenderingBackendCreationParameters::encode const):
(WebKit::RemoteRenderingBackendCreationParameters::decode):
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.h:
(WebKit::InitializationSegmentInfo::TrackInformation::ipcData):
(WebKit::InitializationSegmentInfo::TrackInformation::encode const):
(WebKit::InitializationSegmentInfo::TrackInformation::decode):
(WebKit::InitializationSegmentInfo::ipcData):
(WebKit::InitializationSegmentInfo::encode const):
(WebKit::InitializationSegmentInfo::decode):
* Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
(WebKit::MediaDescriptionInfo::ipcData):
(WebKit::MediaDescriptionInfo::encode const):
(WebKit::MediaDescriptionInfo::decode):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.h:
(WebKit::RemoteMediaPlayerProxyConfiguration::ipcData):
(WebKit::RemoteMediaPlayerProxyConfiguration::encode const):
(WebKit::RemoteMediaPlayerProxyConfiguration::decode):
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.h:
(WebKit::TextTrackPrivateRemoteConfiguration::ipcData):
(WebKit::TextTrackPrivateRemoteConfiguration::encode const):
(WebKit::TextTrackPrivateRemoteConfiguration::decode):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyDataForSpecificFirstParty::encode const):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyDataForSpecificFirstParty::decode):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyData::encode const):
(WebKit::WebResourceLoadStatisticsStore::ThirdPartyData::decode):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
(WebKit::NetworkSessionCreationParameters::ipcData):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeIPCData):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/WebProcess/GPU/media/MediaOverridesForTesting.h:
(WebKit::MediaOverridesForTesting::ipcData):
(WebKit::MediaOverridesForTesting::encode const):
(WebKit::MediaOverridesForTesting::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSessionConfiguration.h:
(WebKit::RemoteAudioSessionConfiguration::ipcData):
(WebKit::RemoteAudioSessionConfiguration::encode const):
(WebKit::RemoteAudioSessionConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMConfiguration.h:
(WebKit::RemoteCDMConfiguration::ipcData):
(WebKit::RemoteCDMConfiguration::encode const):
(WebKit::RemoteCDMConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceConfiguration.h:
(WebKit::RemoteCDMInstanceConfiguration::ipcData):
(WebKit::RemoteCDMInstanceConfiguration::encode const):
(WebKit::RemoteCDMInstanceConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerConfiguration.h:
(WebKit::RemoteMediaPlayerConfiguration::ipcData):
(WebKit::RemoteMediaPlayerConfiguration::encode const):
(WebKit::RemoteMediaPlayerConfiguration::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.h:
(WebKit::RemoteMediaPlayerState::ipcData):
(WebKit::RemoteMediaPlayerState::encode const):
(WebKit::RemoteMediaPlayerState::decode):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
(WebKit::RemoteVideoFrameProxy::Properties::encode const): Deleted.
(WebKit::RemoteVideoFrameProxy::Properties::decode): Deleted.
* Source/WebKit/WebProcess/Network/NetworkProcessConnectionInfo.h:
(WebKit::NetworkProcessConnectionInfo::ipcData):
(WebKit::NetworkProcessConnectionInfo::encode const):
(WebKit::NetworkProcessConnectionInfo::decode):
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.h:
(WebKit::WebSpeechSynthesisVoice::ipcData):
(WebKit::WebSpeechSynthesisVoice::encode const):
(WebKit::WebSpeechSynthesisVoice::decode):
